docs/pipeline_semantics.md

```diff
@@ -0,0 +1,173 @@
+meaning that they need to pass the message to the next node before taking another message¹. This also means there is no
+explicit buffer in Conduit, a pipeline can only hold only as many messages as there are nodes in the pipeline (see
+[backpressure](#Backpressure) for more information).
+
+¹This might change in the future if we decide to add support for multi-message transforms.
+
+### Message states
+
+A message can be in one of 3 states:
+
+- **Open** - all messages start in the open state. This means the message is currently in processing, either by a node
+  or a destination connector. A pipeline won't stop until all messages transition from the open state into one of the
+  other two states.
+- **Acked** - a message was successfully processed and acknowledged. This can be done either by a processor (e.g. it
+  filtered the message out) or by a destination. If a pipeline contains multiple destinations, the message needs to be
+  acknowledged by all destinations before it is marked as acked. Acks are propagated back to the source connector and
+  can be used to advance the position in the source system if applicable.
+- **Nacked** - a message was negatively acknowledged either by a processor (e.g. a transform failed) or by a
+  destination. If a pipeline contains multiple destinations, the message needs to be negatively acknowledged by at least
+  one destination before it is marked as nacked. When a message is nacked the message is passed to
+  the [DLQ](#Dead-letter-queue).
+
+**Important**: if a message gets nacked and the DLQ handler successfully processes the nack (e.g. stores the message in
+the dead letter queue), the source connector will receive an ack as if the message was successfully processed, even
+though Conduit marks it internally as nacked. In other words, the source connector will receive an ack every time
+Conduit handled a message end-to-end and it can be safely discarded from the source.
+
+Pipeline nodes will either leave the message open and send it to the next node for processing or ack/nack it and not
+send it further down the pipeline. If the ack/nack fails, the node will stop running and return an error that will
+consequently stop the whole pipeline. The returned error is stored in the pipeline for further inspection by the user.
+
+### Message state change handlers
+
+A pipeline node can register state change handlers on a message that will be called when the message state changes. This
+is used for example to register handlers that reroute nacked messages to a dead letter queue or to update metrics when a
+message reaches the end of the pipeline.
+
+## Semantics
+
+### Messages are delivered in order
+
+Since messages are passed between nodes in channels and a node only processes one message at a time, it is guaranteed
+that messages will flow through the Conduit pipeline in the same order that was produced by the source.
```

Comment:
>  it is guaranteed that messages will flow 
Just so nobody gets the impression it's messages generally that will flow through a pipeline in the same order, it might be good to say "that messages from a single source connector will flow...". I do see there's a related caveat mention, but I think it's good to be clear here too.